oak/server/rust/oak_runtime/src/node/wasm.rs

```diff
@@ -0,0 +1,785 @@
+            "channel_close" => (
+                CHANNEL_CLOSE,
+                wasmi::Signature::new(&[ABI_U64][..], Some(ABI_U32)),
+            ),
+            //
+            // - 3: channel_create: (usize, usize) -> u32
+            "channel_create" => (
+                CHANNEL_CREATE,
+                wasmi::Signature::new(&[ABI_USIZE, ABI_USIZE][..], Some(ABI_U32)),
+            ),
+            //
+            // - 4: channel_write: (u64, usize, usize, usize, u32) -> u32
+            "channel_write" => (
+                CHANNEL_WRITE,
+                wasmi::Signature::new(
+                    &[ABI_U64, ABI_USIZE, ABI_USIZE, ABI_USIZE, ABI_U32][..],
+                    Some(ABI_U32),
+                ),
+            ),
+            //
+            // - 5: channel_read: (u64, usize, usize, usize, usize, u32, usize) -> u32
+            "channel_read" => (
+                CHANNEL_READ,
+                wasmi::Signature::new(
+                    &[
+                        ABI_U64, ABI_USIZE, ABI_USIZE, ABI_USIZE, ABI_USIZE, ABI_U32, ABI_USIZE,
+                    ][..],
+                    Some(ABI_U32),
+                ),
+            ),
+            //
+            // - 6: wait_on_channels: (usize, u32) -> u32
+            "wait_on_channels" => (
+                WAIT_ON_CHANNELS,
+                wasmi::Signature::new(&[ABI_USIZE, ABI_U32][..], Some(ABI_U32)),
+            ),
+            _ => {
+                return Err(wasmi::Error::Instantiation(format!(
+                    "Export {} not found",
+                    field_name
+                )))
+            }
+        };
+
+        if &sig != signature {
+            return Err(wasmi::Error::Instantiation(format!(
+                "Export `{}` doesnt match expected type {:?}",
+                field_name, signature
+            )));
+        }
+
+        Ok(wasmi::FuncInstance::alloc_host(sig, index))
+    }
+}
+
+/// Create a new instance of a Wasm node. If the entry point is not found,
+/// `ERR(OakStatus::ERR_INVALID_ARGS)` will be returned immediately.
+pub fn new_instance(
+    config_name: &str,
+    runtime: RuntimeRef,
+    module: Arc<wasmi::Module>,
+    entrypoint: String,
+    handle: ChannelReader,
+) -> Result<crate::JoinHandle, OakStatus> {
+    let config_name = config_name.to_owned();
+
+    // wasmi can't enumerate exports at creation, so we have to do it here per instance spawned as
+    // the entrypoint could be anything. We do it before spawning the child thread so that we can
+    // return an error code immediately if appropriate.
+    {
+        let abi = WasmInterface::new(config_name.clone(), runtime.clone(), handle.clone());
+        let instance = wasmi::ModuleInstance::new(
+            &module,
+            &wasmi::ImportsBuilder::new().with_resolver("oak", &abi),
+        )
+        .expect("failed to instantiate wasm module")
+        .assert_no_start();
+
+        if instance
+            .export_by_name(&entrypoint)
```

Comment:
Is there a check somewhere that the entrypoint has the correct signature (cf. [C++](https://github.com/project-oak/oak/blob/master/oak/server/wasm_node.cc#L144-L148)) ?  I guess there must be if the [relevant test](https://github.com/project-oak/oak/blob/master/examples/abitest/module_0/rust/src/lib.rs#L904-L911) is passing but I couldn't immediately see where&hellip;

Reply:
Good catch! This confused me for a moment as there is no good reason my code should have passed this test. Turns out the definition in `module_1` is missing a `#[no_mangle]`. I've added a signature check to `wasm.rs` and also `#[no_mangle]` to the definition.

Reply:
Bah, I keep getting caught out by the `[no_mangle]` thing because `extern "C"` effectively [*means* no_mangle in C++](http://lurklurk.org/linkers/linkers.html#namemangling).  Thanks for fixing.